my/core/query.py

```diff
@@ -0,0 +1,717 @@
+                order_by_chosen = lambda o: order_by_lookup[_determine_order_by_value_key(o)](o)
+
+        assert order_by_chosen is not None
+        # note: can't just attach sort unsortable values in the same iterable as the
+        # other items because they don't have any lookups for order_key or functions
+        # to handle items in the order_by_lookup dictionary
+        unsortable, itr = _handle_unsorted(itr, order_by_chosen, drop_unsorted, wrap_unsorted)
+
+        # run the sort, with the computed order by function
+        itr = iter(sorted(itr, key=order_by_chosen, reverse=reverse))  # type: ignore[arg-type, type-var]
+
+        # re-attach unsortable values to the front/back of the list
+        if reverse:
+            itr = itertools.chain(itr, unsortable)
+        else:
+            itr = itertools.chain(unsortable, itr)
+    else:
+        # if not already done in the order_by block, reverse if specified
+        if reverse:
+            itr = more_itertools.always_reversible(itr)
+
+    # apply limit argument
+    if limit is not None:
+        return itertools.islice(itr, limit)
+
+    return itr
+
+
+timedelta_regex = re.compile(r"^((?P<weeks>[\.\d]+?)w)?((?P<days>[\.\d]+?)d)?((?P<hours>[\.\d]+?)h)?((?P<minutes>[\.\d]+?)m)?((?P<seconds>[\.\d]+?)s)?$")
+
+
+# https://stackoverflow.com/a/51916936
+def parse_timedelta_string(timedelta_str: str) -> timedelta:
+    """
+    This uses a syntax similar to the 'GNU sleep' command
+    e.g.: 1w5d5h10m50s means '1 week, 5 days, 5 hours, 10 minutes, 50 seconds'
+    """
+    parts = timedelta_regex.match(timedelta_str)
+    if parts is None:
+        raise ValueError(f"Could not parse time duration from {timedelta_str}.\nValid examples: '8h', '1w2d8h5m20s', '2m4s'")
+    time_params = {name: float(param) for name, param in parts.groupdict().items() if param}
+    return timedelta(**time_params)  # type: ignore[arg-type]
+
+
+def test_parse_timedelta_string():
+
+    import pytest
+
+    with pytest.raises(ValueError) as v:
+        parse_timedelta_string("5xxx")
+
+    assert v is not None
+    assert str(v.value).startswith("Could not parse time duration from")
+
+    res = parse_timedelta_string("1w5d5h10m50s")
+    assert res == timedelta(days=7.0 + 5.0, hours=5.0, minutes=10.0, seconds=50.0)
+
+
+# classes to use in tests, need to be defined at the top level
+# because of a mypy bug
+class _Int(NamedTuple):
+    x: int
+
+
+# to test order_key with compatible orderable (int, float) types
+class _Float(NamedTuple):
+    x: float
+
+
+def test_basic_orders() -> None:
+
+    import random
+
+    import pytest
+
+    def basic_iter() -> Iterator[_Int]:
+        for v in range(1, 6):
+            yield _Int(v)
+
+    def filter_two(obj: Any) -> bool:
+        return obj.x != 2
+
+    res = list(select(basic_iter(), where=filter_two, reverse=True))
+    assert len(res) == 4
+
+    for (actual, expected) in zip(res, (5, 4, 3, 1)):
```

Comment:
nit, but I usually try asserting on lists instead of iterating and asserting on individual items -- gives better error messages

Reply:
yeah, makes sense